core/eolearn/core/eodata_io.py

```diff
@@ -0,0 +1,260 @@
+file in the root directory of this source tree.
+"""
+import pickle
+import gzip
+import concurrent.futures
+
+import fs
+from fs.tempfs import TempFS
+import numpy as np
+from sentinelhub.os_utils import sys_is_windows
+
+from .constants import FeatureType, FileFormat, OverwritePermission
+from .utilities import FeatureParser
+
+
+def save_eopatch(eopatch, filesystem, patch_location, features=..., overwrite_permission=OverwritePermission.ADD_ONLY,
+                 compress_level=0):
+    """ A utility function used by EOPatch.save method
+    """
+    patch_exists = filesystem.exists(patch_location)
+
+    if overwrite_permission is OverwritePermission.OVERWRITE_PATCH and patch_exists:
+        filesystem.removetree(patch_location)
+        patch_exists = False
+
+    if not patch_exists:
+        filesystem.makedir(patch_location, recreate=True)
+
+    eopatch_features = list(walk_eopatch(eopatch, patch_location, features))
+
+    if overwrite_permission is OverwritePermission.ADD_ONLY or \
+            (sys_is_windows() and overwrite_permission is OverwritePermission.OVERWRITE_FEATURES):
+        fs_features = list(walk_filesystem(filesystem, patch_location, features))
+    else:
+        fs_features = []
+
+    _check_case_matching(eopatch_features, fs_features)
+
+    if overwrite_permission is OverwritePermission.ADD_ONLY:
+        _check_add_only_permission(eopatch_features, fs_features)
+
+    ftype_folder_map = {(ftype, fs.path.dirname(path)) for ftype, _, path in eopatch_features if not ftype.is_meta()}
+    for ftype, folder in ftype_folder_map:
+        if not filesystem.exists(folder):
+            filesystem.makedirs(folder)
+
+    features_to_save = ((FeatureIO(filesystem, path),
+                         eopatch[(ftype, fname)],
+                         FileFormat.NPY if ftype.is_raster() else FileFormat.PICKLE,
+                         compress_level) for ftype, fname, path in eopatch_features)
+
+    with concurrent.futures.ThreadPoolExecutor() as executor:
+        executor.map(lambda params: params[0].save(*params[1:]), features_to_save)
+
+
+def load_eopatch(eopatch, filesystem, patch_location, features=..., lazy_loading=False):
+    """ A utility function used by EOPatch.load method
+    """
+    features = list(walk_filesystem(filesystem, patch_location, features))
+    loading_data = [FeatureIO(filesystem, path) for _, _, path in features]
+
+    if not lazy_loading:
+        with concurrent.futures.ThreadPoolExecutor() as executor:
+            loading_data = executor.map(lambda loader: loader.load(), loading_data)
+
+    for (ftype, fname, _), value in zip(features, loading_data):
+        eopatch[(ftype, fname)] = value
+
+    return eopatch
+
+
+def walk_filesystem(filesystem, patch_location, features=...):
+    """ Recursively reads a patch_location and returns yields tuples of (feature_type, feature_name, file_path)
+    """
+    existing_features = {ftype: {} for ftype in FeatureType}
```

Comment:
existing_features = defaultdict(dict) does what you want here... Not sure if its better tho, might be worth considering.

Reply:
Good point, changed.